src/ddspdrum/torchmodule.py

```diff
@@ -485,3 +485,146 @@ def _forward(self, control_in: T, audio_in: T) -> T:
 
         audio_in = fix_length(audio_in, len(control_in))
         return control_in * audio_in
+
+
+class TorchFIR(TorchSynthModule):
+    """
+    A finite impulse response low-pass filter. Uses convolution with a windowed
+    sinc function.
+
+    Parameters
+    ----------
+
+    cutoff (float)      :   cutoff frequency of low-pass in Hz, must be between 5 and
+                            half the sampling rate. Defaults to 1000Hz.
+    filter_length (int) :   The length of the filter in samples. A longer filter will
+                            result in a steeper filter cutoff. Should be greater than 4.
+                            Defaults to 512 samples.
+    sample_rate (int)   :   Sampling rate to run processing at.
+    """
+
+    def __init__(
+        self,
+        cutoff: float = 1000.0,
+        filter_length: int = 512,
+        sample_rate: int = SAMPLE_RATE,
+    ):
+        super().__init__(sample_rate=sample_rate)
+        self.add_parameters(
+            [
+                TorchParameter(
+                    value=cutoff,
+                    parameter_name="cutoff",
+                    parameter_range=ParameterRange(5.0, sample_rate / 2.0, "log")
+                ),
+                TorchParameter(
+                    value=filter_length,
+                    parameter_name="length",
+                    parameter_range=ParameterRange(4.0, 4096.)
+                )
+            ]
+        )
+
+    def _forward(self, audio_in: T) -> T:
+        """
+        Filter audio samples
+        TODO: Cutoff frequency modulation, if there is an efficient way to do it
+
+        Parameters
+        ----------
+
+        audio (T)  :   audio samples to filter
+        """
+
+        impulse = self.windowed_sinc(self.p("cutoff"), self.p("length"))
+        impulse = torch.reshape(impulse, (1, 1, impulse.size()[0]))
+        audio_resized = torch.reshape(audio_in, (1, 1, audio_in.size()[0]))
+        y = nn.functional.conv1d(audio_resized, impulse, padding=int(self.p("length") / 2))
+        return y[0][0]
+
+    def windowed_sinc(self, cutoff: T, filter_length: T) -> T:
```

Comment:
I don't love the code in this method, but it was challenging to make this differentiable in a way that felt really clean. The big obstacle that I encountered in both of the filters was making the filter length differentiable.

Reply:
Cool, might also be worth comparing to https://github.com/mravanelli/SincNet/blob/788fc3cd42e5c3172c5a8525b3a872eb7205c8f9/dnn_models.py and https://github.com/adefossez/julius

Reply:
Nice, the julius code was helpful for cleaning up the windowed sinc implementation a bit. That FFT based convolution implementation is sweet, would be cool to explore that in the future.